sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/changeFeedMetrics/ChangeFeedMetricsTracker.scala

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.changeFeedMetrics
+
+import com.azure.cosmos.implementation.guava25.collect.EvictingQueue
+import com.azure.cosmos.spark.diagnostics.BasicLoggingTrait
+import com.azure.cosmos.spark.{CosmosConstants, NormalizedRange}
+
+// scalastyle:off underscore.import
+import scala.collection.JavaConverters._
+// scalastyle:on underscore.import
+
+private[cosmos] object ChangeFeedMetricsTracker {
+  private val DefaultMaxHistory: Int = CosmosConstants.ChangeFeedTrackerConfigs.changeFeedMetricsTrackerMaxHistory
+  private val DefaultDecayFactor: Double = CosmosConstants.ChangeFeedTrackerConfigs.changeFeedMetricsTrackerDecayFactor
+
+  def apply(partitionIndex: Long, feedRange: NormalizedRange): ChangeFeedMetricsTracker = {
+    new ChangeFeedMetricsTracker(partitionIndex, feedRange)
+  }
+
+  def apply(partitionIndex: Long, feedRange: NormalizedRange, decayFactor: Double): ChangeFeedMetricsTracker = {
+    new ChangeFeedMetricsTracker(partitionIndex, feedRange, DefaultMaxHistory, decayFactor)
+  }
+}
+
+/**
+ * Tracks metrics for Change Feed operations including partition indices and LSN gaps
+ * with exponential weighting for recent metrics
+ */
+private[cosmos] class ChangeFeedMetricsTracker(
+    private val partitionIndex: Long,
+    private val feedRange: NormalizedRange,
+    private val maxHistory: Int = ChangeFeedMetricsTracker.DefaultMaxHistory,
```

Comment:
maybe also make these editable through env variable

Reply:
I think Spark config would be more appropriate if we think we need to make it configurable (my 2 cents - we can make it configurable if we really find the need to - i seriously doubt it will ever be necessary)